remote-trait-object/src/port/server.rs

```diff
@@ -69,21 +70,63 @@ impl Drop for Server {
 
 fn receiver<H>(handler: Arc<H>, ipc_send: Sender<Packet>, ipc_recv: Receiver<Packet>)
 where
-    H: Handler, {
-    loop {
-        let request = match ipc_recv.recv() {
-            Ok(request) => request,
-            Err(_err) => {
-                // ipc_recv is closed.
-                return
-            }
-        };
-
-        trace!("Packet received in Port Server {}", request);
-        let response = handler.handle(request.view());
-        trace!("Handler result in Port Server {:?}", response);
-        let mut response_packet = Packet::new_response_from_request(request.view());
-        response_packet.append_data(&response);
-        ipc_send.send(response_packet).unwrap();
+    H: Handler + 'static, {
+    let received_packets = Arc::new(Queue::new(100));
+    let joiners = create_handler_threads(handler, ipc_send, Arc::clone(&received_packets));
+
+    while let Ok(request) = ipc_recv.recv() {
+        received_packets.push(request).expect("Queue will close after this loop");
+    }
+    // ipc_recv is closed.
+
+    received_packets.close();
+    for joiner in joiners {
+        joiner.join().unwrap();
+    }
+}
+
+fn create_handler_threads<H>(
+    handler: Arc<H>,
+    ipc_send: Sender<Packet>,
+    received_packets: Arc<Queue<Packet>>,
+) -> Vec<thread::JoinHandle<()>>
+where
+    H: Handler + 'static, {
+    let mut joins = Vec::new();
+
+    fn handler_loop<H: Handler>(handler: Arc<H>, ipc_send: Sender<Packet>, received_packets: Arc<Queue<Packet>>) {
+        loop {
+            // FIXME: use condvar instead of timeout
+            let request = match received_packets.pop(Some(time::Duration::from_millis(100))) {
```

Comment:
I don't understand why we need spin here. Why not just `None` for timeout?

Reply:
You are right. I added this before making queue closeable. We don't need timeout since we can close the queue when we want.

Reply:
I fixed it.